Add tagcn

markcheung opened the pull request:
Added Topology Adaptive Graph Convolutional Networks from this paper https://arxiv.org/abs/1710.10370

rusty1s commented:
Thank you. This is a high-quality code PR :)

1. I think the `GCN` normalization is missing.
2. You can remove the `cached` flag from the operator.

markcheung commented:
Thanks.

1. The code is based on TAGCN paper, which uses the adjacency matrix instead of the Laplacian (based on Algebraic GSP https://arxiv.org/pdf/1210.4752.pdf). I don't think any normalization is applied (but can be helpful).

2. I'm not sure if cache is needed and if so, how to use it. Saving y is different from saving x (different dimension). I think it is easier to just remove it for now.

rusty1s commented:
The linked paper states: "We adopt the normalized adjacency matrix to guarantee that all the eigenvalues of A are inside the unit circle", so IMO we should at least provide a flag to let the user decide :)

markcheung commented:
Sounds good, I agree. I will add that in.

markcheung commented:
Updated.

1. I removed 3 lines from [gcn's normalization function](https://github.com/rusty1s/pytorch_geometric/blob/5ea66a21c3e1f16a3deee8bfe46b40b1123691da/torch_geometric/nn/conv/gcn_conv.py#L70) to not add self-loops.
2. Changed default K (number of hops) to 3 since it gives higher accuracy.

rusty1s commented:
Awesome. Thank you!

Christian-Git01 commented:
Should A have an exponent k in the docs? I think that higher orders of A are used.

![image](https://user-images.githubusercontent.com/52127686/76236549-24487480-622d-11ea-94b4-787f1bf12e6c.png)

rusty1s commented:
Yea, that's a bug in the documentation.

Christian-Git01 commented:
Perfect and thanks for replying so fast and fixing it.

Christian-Git01 commented:
Why do you think that there is no normalization? They define the adjacency matrix at the bottom of page 2:
![image](https://user-images.githubusercontent.com/52127686/95971955-29e4e900-0e12-11eb-8f78-dcf1ebc962c3.png)

and later on they define the normalized adjacency matrix:
![image](https://user-images.githubusercontent.com/52127686/95972249-847e4500-0e12-11eb-8b0f-428b477dfcef.png)
I guess that they keep using the normalized adjacency matrix.
